doc/man1/mkriocon.1
.TH MKRIOCON 1 "2008-08-25" 1.0 "LINX"
.SH NAME
mkriocon \- create LINX RapidIO connections.
.SH SYNOPSIS
.B mkriocon [OPTIONS] <connection>

.SH DESCRIPTION
Create a LINX RapidIO connection to a remote node.

.SH OPTIONS

.IP "-h, --help"
Display help and exit.

.IP "Mandatory options"

.IP "-p, --port=<rio port>"
Specify the remote port number of the node to connect to, e.g. 0.

.IP "-l, --local-port=<rio port>"
Specify the local port number to use, e.g. 0.

.IP "-I, --id=<rio device id>"
Specify the RapidIO device id to connect to, e.g. 1.

.IP "-i, --if=<rio device>"
Specify the RapidIO interface to use, e.g. rio0.

.IP "Optional options"

.IP "-t, --tmo=nn"
The connection supervision timeout specifies the time (in hundreds of msec) to
wait until a connection is considered broken. Default is 5 (yielding a timeout
of 500 msec) and should be sufficient in most systems.
Tmo 0 means to use the default timeout.

.IP "-M, --mtu=nnn"
The MTU (Maximum Transmission Unit) specifies the size in  bytes of the
largets packet that the RapidIO protocol can pass onwards (this includes the
size of the RapidIO header). If not set the MTU is fetched from the interface.
Typically an MTU of 256 bytes is used for RapidIO.

.SH FILES
None.

.SH DIAGNOSTICS
None.

.SH KNOWN BUGS
None.

.SH EXAMPLES
.nf
mkriocon --port=0 --local-port=0 --mbox=1 --id=1 --if=rio0 riocon_A
mkriocon -p 0 -l 0 -m 1 -I 1 -i rio0 riocon_A
.fi

.SH "SEE ALSO"
.BR linx "(7), "
.BR linxcfg "(1), "
.BR linxstat "(1), "
.BR mklink "(1), "
.BR rmriocon "(1)"

.SH AUTHOR
Enea LINX team

.SH COPYRIGHT
Copyright (c) 2006-2007, Enea Software AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
Neither the name of Enea Software AB nor the names of its
contributors may be used to endorse or promote products derived from this
software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


ChangeLog
Version 2.5.1 (2011-07-08)
==========
* The limit of number of sockets per link allowed lowered from 1024 to two.
* Removed a check in LINX kernel module for max signal size.
* Fixes for the multicore header in the ECM and Wireshark patch.
		
Version 2.5.0 (2011-04-26)
==========
* Fixed a bug in linxstat.
* Updated manpage for mkethcon and help info for mkethcon and linxcfg to reflect
  reality.
* libcfg was updated to support cmcl connections.
* Fixed minor bugs in rm*con binaries.
* Fixed build errors for various kernel versions. ipv6 for <2.6.22 and db.c for
  newer ones >2.6.36.
* Fixed some issues in the linx Gateway related to LINXCR-1397
* Added support for ipv6 in the tcp connection manager.
* Improved the performance of intra-process communication with pointer passing.
* Fixed some issues found by coverity.
* Removed some warnings when building linxstat.
* Remove all BUG BUG_ON WARN_ON from linx modules to avoid using __FILE__
  define in the errors. This define causes the linx module to be different
  if compiled in different folder.
* Added fixes in the ECM were a malformed received packet could lead to
  kernel panic.
* Fixed issue with multi-access to LINX sockets from user-space
  applications could lead to kernel panic.
* Added new features to linxstat: print info only for a given huntname
  and display only a specified number of characters from huntname
* Added coreid header
* Added support for 2.5.35 kernels.
* Fixed bug with negative signal filter.
* Fixed bug with missing internal signals in the message trace.
* Added patch file for Ethernet CM to compensate for faulty Ethernet
  drivers that don't allocate enough headroom in skb when using VLAN.
* Added support for 2.6.33 and 2.6.34 kernels.
* Fixed vlan bug reproducing on certain machines
* Fixed bug with receive_w_tmo where tmo=0 resulted in no signal above
  4095 could be received.
* Fixed string handling bug in linxgws.
* Fixed memory allocation issue in linxgws.
* Fixed a bug in linxdisc regarding several interfaces.
* Added ip config in linxnet conf file
* Linxstat now uses a function interface to retrieve info instead of
ioctl.
	
Version 2.4.0 (2010-02-22)
==========	
* Added missing riocm.txt to MANIFEST
* Added RIOCM documentation. net/linx/riocm/riocm.txt
* Updated RELEASE_NOTES for 2.4.0 release.
* Added support for creating links using RapidIO CM connections in linxcfg.

Version 2.4.0-beta (2010-01-15)
==========
* Added a RapidIO connection manager.
* Remove debug printk in device notification handler.

Version 2.3.0 (2009-12-9)
==========
* Fixed build errors for vanilla kernel 2.6.32
* Example conf file for Linxdisc now emphasizes that VLAN interfaces are supported.
* RLNH control data is now always sent from kernel context.
* Altered the keepalive mechanism for the Ethernet CM. The old implementation
  was sensitive to packet loss. Contributed by paull7 @ sourceforge. Note: The
  new implementation is backwards compatible.
* Added support for 2.6.31 kernels.
* Fix LINX_IOCTL_INFO ioctl request for gathering process' receive filters.
* Document LINX_INFO_RECV_QUEUE_2.
* Drop tcp connect packets that doesn't match the current protocol
version.
* Add hardware address size in ethernet connection header.
* Man page for linx_request_new_link and linx_cancel_new_link was missing
* Fixed bug where the request ack bit was not set when the sliding window queue
  is becoming full.
* Fixed memory leaks in the code for create and destroy connections and links.	
* New shared memory connection manager.
* Wireshark plugin now supports LINX TCP Connection manager.
* Wrong RLNH protocol version was listed in documentation.	
	
Version 2.2.0 (2009-03-31)
==========
* New Linxdisc and Linx internal messages patches for Wireshark 1.0.6
* New LINX patches for tcpdump 4.0.0 and libpcap 1.0.0
* Trace of internal LINX messages now use official LINX ethernet type, 0x8911.
* Added new function linx_get_owner(), return the pid of the owner of a LINX endpoint.	
* Added missing files to MANIFEST.
* Example files now have the correct copyright header.
* Fixed oob-issue where the a link in some cases wasn't created.
* Removed a state in the tcp cm state machine.
* Did some major cleanup in tcp cm.
* Removed a kernel function from af_linx.c
* Fixed a bug in tcp cm regarding halt of workqueue when connecting to dead peers
* Linux Gateway Server and Client added to distribution, also gwcmd tool to test Linx Gateway Servers. 
* Removed bug_on's and fixed better error reporting back to user space on failed send.
* Fix bug in linx_info_filter_32
* Fixed alignment issue in linxstat.
* Removed linx_create/destroy_eth/tcp_link functions.
* Avoid calling kmalloc(0) from Linx IPC-layer.

Version 2.1.0 (2008-11-10)
==========	
* LINX now runs on x86 2.6.27 Linux kernel.
* Add support for mips(64).
* Add patches to libpcap and tcpdump configure scripts for Octeon.
* The TCP CM now support OOB messaging.
* Fixes for running w errorchecks enabled.
* Improved garbage collection of sockets in the tcp_cm.
* Minor changes of linxstat output.
* The linxstat program can now tell if a signal in the receive queue is OOB.
* Support for allocating fragmentated skb's is now a LINX kernel module param and no longer
  a comple-time option.
* Fixed bug when building the LINX kernel module with memory fragmentation support on and
  sending large unpaged signals.
* Fixed bug in LINX message trace. If a fragmented message was received, only the first
  fragment was traced.
* Added optimized linx_do_sendmsg call for attach/hunt/tmo/newlink signals.
* Pulled back warning level a bit to avoid complaints from kernel headers.
* Fixed bug that lead to the LINX kernel module couldn't be unloaded.
* linx_skb_copy_bits removed
* linx_ioctl_info_*() functions moved to a seperate file.
* Internal handling of attach/tmo/newlink signals is simplified.
* Updated liblinx to user LINX_IOCTL_SEND and LINX_IOCTL_RECEIVE instead of sendto/recvfrom/sendmsg

linxcfg/mkriocon.c
/*
 * Copyright (c) 2006-2007, Enea Software AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of Enea Software AB nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * File: mkriocon.c - Command to make a Linx RapidIO connection.
 */
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <linx.h>
#include <linxcfg.h>
#include "db_utils.h"
#include "riocm_utils.h"

#define OPT_MTU 263

struct option long_opts[] = {
        {"mbox", 1, NULL, (int)'m'},
        {"mtu", 1, NULL, (int)'M'},
        {"port", 1, NULL, (int)'p'},
        {"local-port", 1, NULL, (int)'l'},
        {"id", 1, NULL, (int)'I'},
        {"tmo", 1, NULL, (int)'t'},
        {"if", 1, NULL, (int)'i'},
        {"help", 0, NULL, (int)'h'},
        {"silent", 0, NULL, (int)'s'},
        {NULL, 0, NULL, 0}
};

static char *usage =
"mkriocon [OPTIONS] <connection>\n"
"Create a Linx RapidIO connection\n"
"OPTIONS:\n"
"    -p, --port=<rio port>\n"
"        Specify the remote port to connect to,\n"
"        e.g. 1.\n"
"        Mandatory option.\n"
"\n"
"    -l, --local-port=<rio port>\n"
"        Specify the local port to connect from,\n"
"        e.g. 1.\n"
"        Mandatory option.\n"
"\n"
"    -M, --mtu=<rio mtu>\n"
"        Specify the mtu to be used. Default is fetched from the interface.\n"
"\n"
"    -m, --mbox=<rio mailbox>\n"
"        Specify the remote mailbox to connect to, e.g 0.\n"
"        Mandatory option.\n"
"\n"
"    -I, --id=<rio device id>\n"
"        Specify the remote device ID to connect to,\n"
"        e.g. 4.\n"
"        Mandatory option.\n"
"\n"
"    -t, --tmo=<interface>\n"
"        Specify the connection heartbeat timeout to use.\n"
"        The values is multiplied by 100 so a value of 5\n"
"        will yield a timeout of 500ms, which also is the\n"
"        default value in the CM\n"
"\n"
"    -i, --if=<interface>\n"
"        Specify the RapidIO interface to use, e.g. rio0.\n"
"        Mandatory option.\n"
"\n"
"Examples:\n"
"    mkriocon --port=1 --local-port=1 --mbox=2 --id=4 --if=rio0 riocon_A\n"
"    mkriocon -p 1 -l 1 -m 2 -I 4 -i rio0 riocon_A\n";

int main(int argc, char *argv[])
{
        int c, silent, status, ver;
	int port_not_set, local_port_not_set, mbox_not_set, id_not_set;
        char *s, version[14];
        union linx_con_arg con;

	port_not_set = 1;
	local_port_not_set = 1;
	mbox_not_set = 1;
	id_not_set = 1;
        memset(&con.rio, 0, sizeof(con.rio));
        silent = 0;
        ver = linx_get_version(version);
        if (ver == -1) {
                printf("WARNING: failed to retrieve LINX version\n");
                ver = 0;
                strcpy(version, "unknown");
        }

        for (c = 0; c != -1;) {
                c = getopt_long(argc, argv, "hl:M:m:I:i:p:t:s",
				long_opts, NULL);
                switch (c) {
                case 'h':
                        printf("Usage: %s\n", usage);
                        return 0;
                case 's':
                        silent = 1;
                        break;
                case 'p':
			con.rio.port = (uint16_t)strtoul(optarg, &s, 0);
			port_not_set = 0;
                        break;
                case 'l':
			con.rio.my_port = (uint16_t)strtoul(optarg, &s, 0);
			local_port_not_set = 0;
                        break;
                case 'i':
                        con.rio.rioif = optarg;
                        break;
                case 'I':
                        con.rio.id = (uint16_t)strtoul(optarg, &s, 0);
			id_not_set = 0;
                        break;
                case 'm':
                        con.rio.mbox = (uint8_t)strtoul(optarg, &s, 0);
			mbox_not_set = 0;
                        break;
                case 'M':
                        con.rio.mtu = (uint16_t)strtoul(optarg, &s, 0);
                        break;
                case 't':
                        con.rio.hb = (uint8_t)strtoul(optarg, &s, 0);
                        break;
                case '?':
                        return 1;
                }
        }
        con.rio.name = argv[optind];

        if (con.rio.name == NULL) {
                printf("mkriocon: no connection specified\n"
                       "Try 'mkriocon --help' for more information\n");
                return 1;
        }
        if (port_not_set) {
                printf("mkriocon: -p, --port is missing\n"
                       "Try 'mkriocon --help' for more information\n");
                return 1;
        }
        if (local_port_not_set) {
                printf("mkriocon: -l, --local-port is missing\n"
                       "Try 'mkriocon --help' for more information\n");
                return 1;
        }
        if (id_not_set) {
                printf("mkriocon: -I, --id is missing\n"
                       "Try 'mkriocon --help' for more information\n");
                return 1;
        }
        if (mbox_not_set) {
                printf("mkriocon: -m, --mbox is missing\n"
                       "Try 'mkriocon --help' for more information\n");
                return 1;
        }
        if (con.rio.rioif == NULL) {
                printf("mkriocon: -i, --if is missing\n"
                       "Try 'mkriocon --help' for more information\n");
                return 1;
        }

        s = NULL;
        status = linx_create_connection(LINX_CON_RIO, &con, &s);
        if (status != 0) {
                printf("mkriocon: couldn't create connection '%s': %s\n",
                       con.rio.name, strerror(status));
                return 1;
        }
        
        if (!silent)
                printf("mkriocon: created connection '%s'.\n"
                       "Now use 'mklink -c %s ...' to create a link.\n",
                       s, s);
        free(s);
        return 0;
}
